internal/cmd/commands/database/migrate.go

```diff
@@ -0,0 +1,311 @@
+		Completion: complete.PredictOr(
+			complete.PredictFiles("*.hcl"),
+			complete.PredictFiles("*.json"),
+		),
+		Usage: `Path to a configuration file containing a "kms" block marked for "config" purpose, to perform decryption of the main configuration file. If not set, will look for such a block in the main configuration file, which has some drawbacks; see the help output for "boundary config encrypt -h" for details.`,
+	})
+
+	f.StringVar(&base.StringVar{
+		Name:       "log-level",
+		Target:     &c.flagLogLevel,
+		EnvVar:     "BOUNDARY_LOG_LEVEL",
+		Completion: complete.PredictSet("trace", "debug", "info", "warn", "err"),
+		Usage: "Log verbosity level. Supported values (in order of more detail to less) are " +
+			"\"trace\", \"debug\", \"info\", \"warn\", and \"err\".",
+	})
+
+	f.StringVar(&base.StringVar{
+		Name:       "log-format",
+		Target:     &c.flagLogFormat,
+		Completion: complete.PredictSet("standard", "json"),
+		Usage:      `Log format. Supported values are "standard" and "json".`,
+	})
+
+	f = set.NewFlagSet("Init options")
+
+	f.BoolVar(&base.BoolVar{
+		Name:   "allow-development-migrations",
+		Target: &c.flagAllowDevMigrations,
+		Usage:  "If set the migrate command will continue even if the schema includes database update steps that may not be supported in the next official release.  Boundary does not provide a rollback mechanism so a backup should be taken independently if needed.",
+	})
+
+	f.StringVar(&base.StringVar{
+		Name:   "migration-url",
+		Target: &c.flagMigrationUrl,
+		Usage:  `If set, overrides a migration URL set in config, and specifies the URL used to connect to the database for migration. This can allow different permissions for the user running initialization or migration vs. normal operation. This can refer to a file on disk (file://) from which a URL will be read; an env var (env://) from which the URL will be read; or a direct database URL.`,
+	})
+
+	return set
+}
+
+func (c *MigrateCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictNothing
+}
+
+func (c *MigrateCommand) AutocompleteFlags() complete.Flags {
+	return c.Flags().Completions()
+}
+
+func (c *MigrateCommand) Run(args []string) (retCode int) {
+	if result := c.ParseFlagsAndConfig(args); result > 0 {
+		return result
+	}
+
+	if c.configWrapper != nil {
+		defer func() {
+			if err := c.configWrapper.Finalize(c.Context); err != nil {
+				c.UI.Warn(fmt.Errorf("Error finalizing config kms: %w", err).Error())
+			}
+		}()
+	}
+
+	dialect := "postgres"
+
+	if schema.DevMigration(dialect) != c.flagAllowDevMigrations {
+		if schema.DevMigration(dialect) {
+			c.UI.Error(base.WrapAtLength("This version of the binary has " +
+				"dev database schema updates which may not be supported in the " +
+				"next official release. To proceed anyways please use the " +
+				"'-allow-development-migrations' flag."))
+			return 2
+		} else {
+			c.UI.Error(base.WrapAtLength("The '-allow-development-migrations' " +
```

Comment:
I don't think this should be an error. A user running dev mode may always want to specify that flag just in case, and we shouldn't error out just because there happen to be none at this time.

Reply:
I worry that this flag would end up seen as a "just-make-it-work" flag which ends up being placed everywhere, including in production, where it doesn't belong.  I agree that when there isn't a dev schema change and they've built from head it could get annoying to have to unset this flag, but I'm weighing it against the potential for data loss in production.

Reply:
Wouldn't you not have development migrations in production though? I agree it might just end up being used all over, however, it _ought_ to be a no-op in prod. And in many cases people will indeed want it every time they are running in dev cause they'll start with a known base DB.

Reply:
We discussed over slack that this can be a warning instead.